core/communication_manager.py

```diff
@@ -0,0 +1,155 @@
+        """
+        Method called by other modules to inform the Communication Manager about
+        events that are going on in the service.
+
+        If there isn't an active optimizer, the commmgr will handle it
+
+        Else, payloads are relayed to the active optimizer which decides how
+        and to whom the Communication Manager should communicate the event/message.
+
+        For example: A runner could inform the Communication Manager that the
+        node is done training a particular model, to which the Optimizer could
+        decide it's time to communicate the new weights to the network.
+        """
+        logging.info("Information has been received: type({})".format(event_type))
+        if self.optimizer:
+            # We have an active session.
+            event_type, payload = self.optimizer.ask(event_type, payload)
+            callback = callback_handler(
+                event_type,
+                self.ACTIONABLE_EVENT_TYPE_2_CALLBACK,
+                ActionableEventTypes.NOTHING.value
+            )
+        else:
+            # We need to start a new session.
+            callback = callback_handler(
+                event_type,
+                self.RAW_EVENT_TYPE_2_CALLBACK,
+                RawEventTypes.NOTHING.value
+            )
+        callback(payload)
+
+    # Callbacks
+
+    def _create_session(self, payload):
+        """
+        Creates a new session and optimizer based on the parameters sent by a
+        model developer. It then asks the optimizer what's the first thing to do
+        and the service starts working on that session from that point on.
+        """
+        # NOTE: We removed the 'optimizer_type' argument since for the MVP we're
+        # only considering the 'FederatedAveragingOptimizer' for now.
+        # TODO: We need to incorporate session id's when we're ready.
+        logging.info("New optimizer session is being set up...")
+        content = payload.get('content')
+        self.optimizer = FederatedAveragingOptimizer(content)
+        logging.info("Optimizer session is set! Now doing kickoff...")
+        event_type, payload = self.optimizer.kickoff()
+        callback = callback_handler(
+            event_type,
+            self.ACTIONABLE_EVENT_TYPE_2_CALLBACK,
+            ActionableEventTypes.NOTHING.value
+        )
+        callback(payload)
+        logging.info("Kickoff complete.")
+
+    def _transform_and_split(self, payload):
+        """
+        Transforms and splits the local dataset.
+
+        TO BE IMPLEMENTED.
+        """
+        logging.info("Transforming and splitting... (NOT IMPLEMENTED YET)")
+        pass
+
+    def _schedule_job(self, dmljob_obj):
+        """
+        Helper function that schedules a DML Job through the Scheduler.
+        """
+        if not self.scheduler:
+            raise Exception("Communication Manager needs to be configured first!")
+        if not dmljob_obj or not isinstance(dmljob_obj, DMLJob):
```

Comment:
Redundant, can just do `not isinstance(dmljob_obj, DMLJob)`

Reply:
True

Reply:
@georgymh then why did you add that line when what i had was fine lmao

Reply:
I meant to remove `not dmljob_obj` and use `not isinstance(dmljob_obj, DMLJob)`

Reply:
It's okay, it's defensive and possibly helps us track the bug to the source. Something that could happen is that we could repurpose this routine to do something else than submitting to the Scheduler, in which case having had this check here helps us identify bugs earlier.

Reply:
@kiddyboots216 you kinda hijacked this thread so I'll let you resolve this one